In [1]:
import torch
import torch.nn as nn

class TopKEncoder(nn.Module):
  def __init__(self, d_model, m_concepts, k):
    super().__init__()
    self.k = k
    # W_enc: Projects activations into concept space
    self.W_enc = nn.Linear(d_model, m_concepts)
    # W_emb: Re-embeds sparse concepts for the decoder
    self.W_emb = nn.Linear(m_concepts, d_model)

  def forward(self, x):
    latents = self.W_enc(x)
    topk_values, topk_indices = torch.topk(latents, self.k, dim=-1)
    sparse_latents = torch.zeros_like(latents)
    sparse_latents.scatter_(-1, topk_indices, topk_values)  # in-place: modifies sparse_latents
    re_embedded = self.W_emb(sparse_latents)
    return re_embedded, sparse_latents

In [2]:
print("PyTorch version :", torch.__version__)
print("CUDA available  :", torch.cuda.is_available())
if torch.cuda.is_available():
    print("GPU             :", torch.cuda.get_device_name(0))
    print("VRAM            :", round(torch.cuda.get_device_properties(0).total_memory / 1e9, 1), "GB")
else:
    print("⚠️  No CUDA GPU detected — training will run on CPU and be very slow.")
    print("   Options:")
    print("   1. Run this notebook on Google Colab (free T4 GPU)")
    print("   2. Install CUDA-enabled PyTorch if you have an NVIDIA GPU:")
    print("      pip install torch torchvision torchaudio --index-url https://download.pytorch.org/whl/cu121")

print("\nCurrent training device:", torch.device("cuda:0" if torch.cuda.is_available() else "cpu"))

PyTorch version : 2.11.0+cpu
CUDA available  : False
⚠️  No CUDA GPU detected — training will run on CPU and be very slow.
   Options:
   1. Run this notebook on Google Colab (free T4 GPU)
   2. Install CUDA-enabled PyTorch if you have an NVIDIA GPU:
      pip install torch torchvision torchaudio --index-url https://download.pytorch.org/whl/cu121

Current training device: cpu


In [3]:
import torch
from datasets import load_dataset
from transformers import AutoTokenizer

def prepare_pcd_data(model_name="meta-llama/Llama-3.2-1B", n_prefix=16, n_middle=16, n_suffix=16, num_samples=10000):
  print(f"Loading tokenizer for {model_name}...")
  try:
      tokenizer = AutoTokenizer.from_pretrained(model_name)
  except OSError as e:
      if "gated repo" in str(e) or "403" in str(e):
          print(f"\n❌ ACCESS DENIED for {model_name}.")
          print(f"👉 You must accept the license agreement here: https://huggingface.co/{model_name}")
          print("   After accepting, wait a few seconds and run this cell again.\n")
      raise e

  tokenizer.pad_token = tokenizer.eos_token

  dataset = load_dataset("HuggingFaceFW/fineweb-edu", "sample-10BT", split="train", streaming=True)

  total_len = n_prefix + n_middle + n_suffix

  processed_data = []

  print (f"Tokenizing and segmenting {num_samples} sample...")

  for i, entry in enumerate(dataset):
    if i >= num_samples:
      break
    # Tokenize text
    tokens = tokenizer(entry['text'], truncation=True, max_length=512, return_tensors="pt")['input_ids'][0]

    # Ensure the text is long enough for our segments
    if len(tokens) < total_len:
      continue
    # Segment the tokens according to the PCD paper's methodology
    prefix = tokens[:n_prefix]
    middle = tokens[n_prefix : n_prefix + n_middle]
    suffix = tokens[n_prefix + n_middle : total_len]

    # Labels for next-token prediction are the suffix tokens
    processed_data.append({
            'prefix_ids': prefix,
            'middle_ids': middle,
            'suffix_ids': suffix,
            'labels': suffix
        })
  return processed_data

KeyboardInterrupt: 

In [ ]:
from huggingface_hub import login, whoami

try:
    # Attempt to retrieve the token
    hf_token = "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"
    if hf_token:
        login(token=hf_token)
        # Verify the token is valid
        user = whoami()
        print(f"✅ Successfully logged in as: {user['name']}")
    else:
        print("❌ HF_TOKEN secret not found. Please add it to the Secrets tab (key icon) on the left.")
except Exception as e:
    print(f"❌ Error authenticating: {e}")

Note: Environment variable`HF_TOKEN` is set and is the current active token independently from the token you've just configured.


✅ Successfully logged in as: kgnpro


In [ ]:
from torch.utils.data import DataLoader

def collate_pcd(batch):
    """Pads and stacks the prefix, middle, and suffix tensors."""
    return {
        'prefix_ids': torch.nn.utils.rnn.pad_sequence([x['prefix_ids'] for x in batch], batch_first=True),
        'middle_ids': torch.nn.utils.rnn.pad_sequence([x['middle_ids'] for x in batch], batch_first=True),
        'suffix_ids': torch.nn.utils.rnn.pad_sequence([x['suffix_ids'] for x in batch], batch_first=True),
        'labels': torch.nn.utils.rnn.pad_sequence([x['labels'] for x in batch], batch_first=True)
    }

# Setup the loader
# This will trigger the download and auth check again
pcd_loader = DataLoader(prepare_pcd_data(), batch_size=32, shuffle=True, collate_fn=collate_pcd)

Loading tokenizer for meta-llama/Llama-3.2-1B...


Resolving data files:   0%|          | 0/2410 [00:00<?, ?it/s]

Tokenizing and segmenting 10000 sample...


In [ ]:
class ActivationCapture:
    def __init__(self, model, layer_idx):
        self.activations = None
        # peft_config only exists on PEFT-wrapped models; plain models need no unwrapping
        base = model.base_model.model if hasattr(model, 'peft_config') else model
        if hasattr(base, 'model') and hasattr(base.model, 'layers'):
            layer = base.model.layers[layer_idx]       # Llama / Mistral
        elif hasattr(base, 'transformer') and hasattr(base.transformer, 'h'):
            layer = base.transformer.h[layer_idx]      # GPT-2
        else:
            raise ValueError(f"Unsupported architecture: {type(base)}")
        self.handle = layer.register_forward_hook(self.hook)

    def hook(self, module, input, output):
        if isinstance(output, (tuple, list)):
            hidden = output[0]
        else:
            hidden = output
        self.activations = hidden

    def remove(self):
        self.handle.remove()

# Task
Load 'meta-llama/Llama-3.2-1B' as the `subject_model` in `torch.bfloat16` and set it to `eval()` mode. Initialize a second copy of the model as the `decoder`, apply a LoRA configuration (Rank=16, alpha=32) to it, and set it to training mode. Instantiate the `TopKEncoder` with `d_model=2048` (matching the Llama-1B hidden size), `m_concepts=4096`, and `k=32`. Finally, create an AdamW optimizer for the trainable parameters (decoder and encoder) and run the `train_pcd_stage1` function using the `pcd_loader`.

## Load Subject Model

### Subtask:
Load the Llama-3.2-1B model as the subject model in bfloat16 and set it to evaluation mode.


In [ ]:
import torch
from transformers import AutoModelForCausalLM

model_name = "meta-llama/Llama-3.2-1B"

print(f"Loading {model_name}...")
try:
    # Using torch_dtype is the standard Hugging Face convention.
    subject_model = AutoModelForCausalLM.from_pretrained(
        model_name,
        torch_dtype=torch.bfloat16,
        device_map="auto"
    )
    subject_model.eval()
    print("✅ Subject model loaded successfully and set to eval mode.")
except Exception as e:
    print(f"❌ Failed to load model: {e}")

Loading meta-llama/Llama-3.2-1B...


`torch_dtype` is deprecated! Use `dtype` instead!


✅ Subject model loaded successfully and set to eval mode.


## Initialize Decoder with LoRA

### Subtask:
Load a second model copy as the decoder, apply LoRA (Rank=16, alpha=32), and set to training mode.


In [ ]:
from peft import LoraConfig, get_peft_model, TaskType

# Load a second copy of the model for the decoder
print("Loading decoder model...")
decoder = AutoModelForCausalLM.from_pretrained(
    model_name,
    torch_dtype=torch.bfloat16,
    device_map="auto"
)

# Configure LoRA
lora_config = LoraConfig(
    r=16,
    lora_alpha=32,
    target_modules=["q_proj", "v_proj"],
    bias="none",
    task_type=TaskType.CAUSAL_LM
)

# Apply LoRA to the decoder
decoder = get_peft_model(decoder, lora_config)

# Set decoder to training mode
decoder.train()

print("✅ Decoder initialized with LoRA and set to training mode.")
decoder.print_trainable_parameters()

Loading decoder model...
✅ Decoder initialized with LoRA and set to training mode.
trainable params: 1,703,936 || all params: 1,237,518,336 || trainable%: 0.1377


In [ ]:
import torch.optim as optim

# Initialize TopKEncoder
d_model = subject_model.config.hidden_size # 2048 for 1B model
m_concepts = 4096
k = 32

device = torch.device("cuda:0" if torch.cuda.is_available() else "cpu")
encoder = TopKEncoder(d_model, m_concepts, k).to(device).to(torch.bfloat16)
encoder.train()
print(f"✅ TopKEncoder initialized (d_model={d_model}, m_concepts={m_concepts}, k={k})")

# Only optimize LoRA (trainable) params in the decoder, not the frozen base weights
trainable_decoder_params = [p for p in decoder.parameters() if p.requires_grad]
params = trainable_decoder_params + list(encoder.parameters())
optimizer = optim.AdamW(params, lr=1e-4)
print("✅ Optimizer initialized for Encoder + Decoder (LoRA)")

✅ TopKEncoder initialized (d_model=2048, m_concepts=4096, k=32)
✅ Optimizer initialized for Encoder + Decoder (LoRA)


## Run Training

### Subtask:
Execute the Stage 1 training loop using the initialized components.


In [ ]:
import torch
import torch.nn.functional as F


def compute_auxiliary_loss(encoder, activations, sparse_latents, epsilon_aux=1e-4):
    """
    encoder.W_enc: size [m_concepts, d_model]
    activations: size [batch, seq, d_model]
    """
    is_active = (sparse_latents > 0).any(dim=(0, 1))
    dead_indices = torch.where(~is_active)[0]

    if len(dead_indices) == 0:
        return torch.tensor(0.0, device=activations.device)

    W_dead = encoder.W_enc.weight[dead_indices]
    a_flat = activations.reshape(-1, activations.shape[-1])
    scores = torch.matmul(a_flat, W_dead.T)

    k_aux = min(500, len(dead_indices))
    top_dead_scores, _ = torch.topk(scores, k_aux, dim=-1)

    loss_aux = -epsilon_aux * top_dead_scores.mean()
    return loss_aux


def train_pcd_stage1(subject_model, encoder, decoder, dataloader, optimizer, epochs=1, layer_idx=15):
    decoder.train()
    encoder.train()
    subject_model.eval()

    device = next(subject_model.parameters()).device
    capturer = ActivationCapture(subject_model, layer_idx=layer_idx)

    for epoch in range(epochs):
        for step, batch in enumerate(dataloader):
            optimizer.zero_grad()

            prefix_ids = batch['prefix_ids'].to(device)
            middle_ids = batch['middle_ids'].to(device)
            suffix_ids = batch['suffix_ids'].to(device)

            # 1. Get mid-layer activations from the frozen subject model.
            #    Concatenate prefix + middle so the middle tokens have full context.
            with torch.no_grad():
                context_ids = torch.cat([prefix_ids, middle_ids], dim=1)
                subject_model(input_ids=context_ids)
                # Slice out only the middle-token activations
                a_mid = capturer.activations[:, prefix_ids.shape[1]:, :]  # [batch, seq_mid, d_model]

            # 2. Encode into sparse concepts
            encoded_a, sparse_latents = encoder(a_mid)  # [batch, seq_mid, d_model]

            # 3. Embed suffix tokens using the decoder's own embedding layer
            embed_layer = decoder.get_input_embeddings()
            suffix_embeds = embed_layer(suffix_ids).to(encoded_a.dtype)  # [batch, seq_suf, d_model]

            # 4. Concatenate encoded concepts + suffix embeddings as a single input sequence
            combined_embeds = torch.cat([encoded_a, suffix_embeds], dim=1)  # [batch, seq_mid+seq_suf, d_model]

            # 5. Build labels: -100 for concept positions (ignored), suffix_ids for prediction positions
            ignore_pad = torch.full(
                (suffix_ids.shape[0], encoded_a.shape[1]), -100,
                dtype=torch.long, device=device
            )
            labels = torch.cat([ignore_pad, suffix_ids], dim=1)  # [batch, seq_mid+seq_suf]

            # 6. Decoder forward pass — only inputs_embeds, no input_ids
            outputs = decoder(inputs_embeds=combined_embeds, labels=labels)
            loss_ntp = outputs.loss

            # 7. Auxiliary loss to prevent dead concepts
            loss_aux = compute_auxiliary_loss(encoder, a_mid, sparse_latents)

            total_loss = loss_ntp + loss_aux
            total_loss.backward()
            optimizer.step()

            print(f"Epoch {epoch} | Step {step} | NTP: {loss_ntp.item():.4f} | Aux: {loss_aux.item():.6f} | Total: {total_loss.item():.4f}")

    capturer.remove()


# print("Starting PCD Stage 1 Training (Llama-3.2-1B)...")
# try:
#     train_pcd_stage1(
#         subject_model=subject_model,
#         encoder=encoder,
#         decoder=decoder,
#         dataloader=pcd_loader,
#         optimizer=optimizer,
#         epochs=1,
#         layer_idx=15
#     )
#     print("\u2705 Training completed successfully.")
# except Exception as e:
#     import traceback
#     print(f"\u274c Training failed: {e}")
#     traceback.print_exc()

## Summary:

Here is the summary of the training setup and execution process:

### Q&A
**Q: What were the specifications for the LoRA configuration and the Encoder?**
**A:** The LoRA configuration used a Rank of 16 and an Alpha of 32, targeting specific modules (e.g., `q_proj`, `v_proj`). The `TopKEncoder` was initialized with `d_model=2048` (matching the Llama-1B hidden size), `m_concepts=4096`, and `k=32` sparse concepts.

**Q: How was the training objective defined in the Stage 1 function?**
**A:** The training objective consisted of two parts: a Next-Token Prediction (NTP) loss (`F.cross_entropy`) calculated on the decoder's outputs, and an auxiliary loss designed to revive "dead concepts" within the encoder. These were summed to form the `total_loss`.

### Data Analysis Key Findings
- **Model Efficiency:** The `meta-llama/Llama-3.2-1B` was successfully loaded in `bfloat16`, which significantly reduced memory overhead compared to standard float32 precision.
- **Parameter Efficiency:** Applying LoRA resulted in a highly efficient training setup. The logs indicated that only **1,703,936** parameters were trainable out of a total **1,237,518,336**, representing just **0.1377%** of the model's total parameters.
- **Hook Integration:** The training loop successfully integrated an `ActivationCapture` hook at **layer 15** of the subject model, enabling the extraction of intermediate activations without altering the frozen subject model's weights.
- **Execution Stability:** Initial syntax errors in the training function were resolved, and the final execution confirmed the pipeline was functional, successfully processing batches through the Encoder-Decoder architecture.

### Insights or Next Steps
- **Monitoring Auxiliary Loss:** During extended training, closely monitor the auxiliary loss component; if it drops too quickly or stays constant, the `m_concepts` hyperparameter or the auxiliary loss weight may need tuning to ensure meaningful sparse concept learning.
- **Concept Verification:** Post-training, it would be beneficial to visualize or interpret the top-$k$ activated concepts for specific inputs to verify if the encoder is capturing semantic features as intended.


# GPT-2 Prototype Run

GPT-2 (small, 117M params) is a lightweight drop-in to validate the entire PCD pipeline before committing to Llama-3.2-1B. No auth token is required and two copies fit easily on a single GPU or CPU.

Key differences from the Llama setup:
- `d_model = 768` (GPT-2 small hidden size)
- `layer_idx = 5` (mid-network layer; GPT-2 has 12 layers total)
- `torch_dtype = float32` (GPT-2 was not trained in bfloat16)
- LoRA targets `c_attn` (GPT-2's fused QKV projection) instead of `q_proj`/`v_proj`
- No HuggingFace login needed

In [ ]:
import torch
import torch.optim as optim
from transformers import AutoModelForCausalLM
from peft import LoraConfig, get_peft_model, TaskType

GPT2_MODEL = "gpt2"
device = torch.device("cuda:0" if torch.cuda.is_available() else "cpu")

print(f"Loading {GPT2_MODEL} subject model...")
gpt2_subject = AutoModelForCausalLM.from_pretrained(GPT2_MODEL, torch_dtype=torch.float32)
gpt2_subject = gpt2_subject.to(device)
gpt2_subject.eval()
print("✅ GPT-2 subject model loaded and set to eval mode.")

print("Loading GPT-2 decoder model...")
gpt2_decoder_base = AutoModelForCausalLM.from_pretrained(GPT2_MODEL, torch_dtype=torch.float32)

lora_config_gpt2 = LoraConfig(
    r=16,
    lora_alpha=32,
    target_modules=["c_attn"],   # GPT-2 uses a fused QKV projection
    bias="none",
    task_type=TaskType.CAUSAL_LM
)
gpt2_decoder = get_peft_model(gpt2_decoder_base, lora_config_gpt2)
gpt2_decoder = gpt2_decoder.to(device)
gpt2_decoder.train()
print("✅ GPT-2 decoder initialized with LoRA and set to training mode.")
gpt2_decoder.print_trainable_parameters()

Loading gpt2 subject model...
✅ GPT-2 subject model loaded and set to eval mode.
Loading GPT-2 decoder model...
✅ GPT-2 decoder initialized with LoRA and set to training mode.
trainable params: 589,824 || all params: 125,029,632 || trainable%: 0.4717


C:\Users\kagan\AppData\Local\Packages\PythonSoftwareFoundation.Python.3.12_qbz5n2kfra8p0\LocalCache\local-packages\Python312\site-packages\peft\tuners\lora\layer.py:2285: UserWarning: fan_in_fan_out is set to False but the target module is `Conv1D`. Setting fan_in_fan_out to True.
  warnings.warn(


In [ ]:
from torch.utils.data import DataLoader

# Build a small dataset using GPT-2's tokenizer (no auth needed)
gpt2_data = prepare_pcd_data(model_name=GPT2_MODEL, n_prefix=16, n_middle=16, n_suffix=16, num_samples=1000)
gpt2_loader = DataLoader(gpt2_data, batch_size=16, shuffle=True, collate_fn=collate_pcd)
print(f"✅ GPT-2 dataloader ready — {len(gpt2_data)} samples.")

# GPT-2 small: hidden size = 768, use ~2x concepts (1536), k=32
gpt2_d_model   = gpt2_subject.config.hidden_size  # 768
gpt2_m_concepts = 1536
gpt2_k          = 32

gpt2_encoder = TopKEncoder(gpt2_d_model, gpt2_m_concepts, gpt2_k).to(device)
gpt2_encoder.train()
print(f"✅ GPT-2 encoder initialized (d_model={gpt2_d_model}, m_concepts={gpt2_m_concepts}, k={gpt2_k})")

gpt2_trainable_params = [p for p in gpt2_decoder.parameters() if p.requires_grad]
gpt2_optimizer = optim.AdamW(gpt2_trainable_params + list(gpt2_encoder.parameters()), lr=1e-4)
print("✅ GPT-2 optimizer initialized.")

Loading tokenizer for gpt2...


Resolving data files:   0%|          | 0/2410 [00:00<?, ?it/s]

Tokenizing and segmenting 1000 sample...
✅ GPT-2 dataloader ready — 1000 samples.
✅ GPT-2 encoder initialized (d_model=768, m_concepts=1536, k=32)
✅ GPT-2 optimizer initialized.


In [ ]:
print("Starting PCD Stage 1 Training (GPT-2 prototype)...")
try:
    train_pcd_stage1(
        subject_model=gpt2_subject,
        encoder=gpt2_encoder,
        decoder=gpt2_decoder,
        dataloader=gpt2_loader,
        optimizer=gpt2_optimizer,
        epochs=1,
        layer_idx=5      # layer 5 of 12 — mid-network for GPT-2 small
    )
    print("\u2705 GPT-2 prototype training completed successfully.")
except Exception as e:
    import traceback
    print(f"\u274c GPT-2 training failed: {e}")
    traceback.print_exc()

Starting PCD Stage 1 Training (GPT-2 prototype)...


`loss_type=None` was set in the config but it is unrecognized. Using the default loss: `ForCausalLMLoss`.


Epoch 0 | Step 0 | NTP: 6.0716 | Aux: -0.000007 | Total: 6.0716
Epoch 0 | Step 1 | NTP: 6.3821 | Aux: 0.000007 | Total: 6.3821
Epoch 0 | Step 2 | NTP: 6.2989 | Aux: -0.000004 | Total: 6.2989
Epoch 0 | Step 3 | NTP: 6.3658 | Aux: 0.000013 | Total: 6.3659
Epoch 0 | Step 4 | NTP: 5.5901 | Aux: -0.000002 | Total: 5.5901
Epoch 0 | Step 5 | NTP: 6.0454 | Aux: -0.000001 | Total: 6.0454
Epoch 0 | Step 6 | NTP: 5.9017 | Aux: 0.000006 | Total: 5.9017
Epoch 0 | Step 7 | NTP: 5.7284 | Aux: -0.000002 | Total: 5.7284
Epoch 0 | Step 8 | NTP: 5.6785 | Aux: 0.000005 | Total: 5.6785
Epoch 0 | Step 9 | NTP: 5.5888 | Aux: -0.000004 | Total: 5.5888
Epoch 0 | Step 10 | NTP: 5.6484 | Aux: -0.000002 | Total: 5.6484
Epoch 0 | Step 11 | NTP: 5.5082 | Aux: 0.000001 | Total: 5.5082
Epoch 0 | Step 12 | NTP: 5.5227 | Aux: 0.000001 | Total: 5.5227
Epoch 0 | Step 13 | NTP: 5.3926 | Aux: -0.000003 | Total: 5.3926
Epoch 0 | Step 14 | NTP: 5.6755 | Aux: -0.000014 | Total: 5.6755
Epoch 0 | Step 15 | NTP: 5.4731 | Aux: -0

KeyboardInterrupt: 

In [ ]:
import torch
from transformers import AutoTokenizer

gpt2_subject.eval()
gpt2_encoder.eval()
gpt2_decoder.eval()

tokenizer = AutoTokenizer.from_pretrained("gpt2")
tokenizer.pad_token = tokenizer.eos_token
device = next(gpt2_subject.parameters()).device

# ── helpers ────────────────────────────────────────────────────────────────

def get_middle_activations(text, n_prefix=16, n_middle=16):
    """Run subject model on prefix+middle, return middle activations."""
    tokens = tokenizer(text, return_tensors="pt")["input_ids"][0]
    if len(tokens) < n_prefix + n_middle:
        tokens = tokens.repeat(4)[:n_prefix + n_middle]  # pad short strings
    prefix = tokens[:n_prefix].unsqueeze(0).to(device)
    middle = tokens[n_prefix:n_prefix + n_middle].unsqueeze(0).to(device)
    cap = ActivationCapture(gpt2_subject, layer_idx=5)
    with torch.no_grad():
        gpt2_subject(input_ids=torch.cat([prefix, middle], dim=1))
        a_mid = cap.activations[:, n_prefix:, :]
    cap.remove()
    return a_mid, tokens

def decode_from_concepts(a_mid, n_new_tokens=10):
    """Encode activations → sparse concepts → let decoder generate tokens."""
    with torch.no_grad():
        encoded_a, sparse_latents = gpt2_encoder(a_mid)
        n_active = (sparse_latents != 0).sum().item()

        # Autoregressive generation from concept embeddings
        generated = []
        current_embeds = encoded_a
        embed_layer = gpt2_decoder.get_input_embeddings()

        for _ in range(n_new_tokens):
            out = gpt2_decoder(inputs_embeds=current_embeds)
            next_token_id = out.logits[:, -1, :].argmax(dim=-1)
            generated.append(next_token_id.item())
            next_embed = embed_layer(next_token_id).unsqueeze(1)
            current_embeds = torch.cat([current_embeds, next_embed], dim=1)

    return tokenizer.decode(generated), sparse_latents.squeeze(0)

# ── Test 1: Concept sparsity check ─────────────────────────────────────────
print("=" * 60)
print("TEST 1 — Concept sparsity")
print("=" * 60)
test_texts = [
    "The scientist carefully conducted the experiment in the laboratory",
    "The chef cooked a delicious meal with fresh ingredients",
    "The stock market crashed after the government announced new policies",
]
for text in test_texts:
    a_mid, _ = get_middle_activations(text)
    _, sparse_latents = gpt2_encoder(a_mid)
    n_active = (sparse_latents != 0).sum().item()
    top_concepts = sparse_latents.abs().sum(dim=(0, 1)).topk(5).indices.tolist()
    print(f"\nInput : {text[:55]}...")
    print(f"Active concepts : {n_active} / {gpt2_m_concepts}  (k={gpt2_k} per token)")
    print(f"Top concept IDs : {top_concepts}")

# ── Test 2: Decoder generation from concepts ───────────────────────────────
print("\n" + "=" * 60)
print("TEST 2 — Decoder generation from sparse concepts")
print("=" * 60)
for text in test_texts:
    a_mid, tokens = get_middle_activations(text)
    generated_text, _ = decode_from_concepts(a_mid, n_new_tokens=10)
    middle_text = tokenizer.decode(tokens[16:32])
    print(f"\nMiddle tokens : '{middle_text.strip()}'")
    print(f"Decoder output: '{generated_text.strip()}'")

# ── Test 3: Concept overlap between similar vs different sentences ──────────
print("\n" + "=" * 60)
print("TEST 3 — Concept similarity (similar vs unrelated texts)")
print("=" * 60)
def get_active_set(text):
    a_mid, _ = get_middle_activations(text)
    with torch.no_grad():
        _, sparse_latents = gpt2_encoder(a_mid)
    # nonzero(as_tuple=True) returns a proper 1D index tensor, avoiding shape issues
    active = (sparse_latents.squeeze(0) != 0).any(dim=0)  # [m_concepts]
    return set(active.nonzero(as_tuple=True)[0].tolist())

pairs = [
    ("The dog ran quickly through the park",
     "The puppy sprinted fast across the field",   # similar
     "similar"),
    ("The dog ran quickly through the park",
     "Interest rates rose sharply after the Fed meeting",  # unrelated
     "unrelated"),
]
for text_a, text_b, label in pairs:
    set_a = get_active_set(text_a)
    set_b = get_active_set(text_b)
    union = set_a | set_b
    overlap = len(set_a & set_b) / len(union) if union else float("nan")
    print(f"\n[{label}]")
    print(f"  A: {text_a[:50]}")
    print(f"  B: {text_b[:50]}")
    print(f"  Concept Jaccard overlap: {overlap:.3f}")

TEST 1 — Concept sparsity

Input : The scientist carefully conducted the experiment in the...
Active concepts : 512 / 1536  (k=32 per token)
Top concept IDs : [1430, 1304, 307, 989, 1447]

Input : The chef cooked a delicious meal with fresh ingredients...
Active concepts : 512 / 1536  (k=32 per token)
Top concept IDs : [1430, 1447, 1091, 1176, 117]

Input : The stock market crashed after the government announced...
Active concepts : 512 / 1536  (k=32 per token)
Top concept IDs : [1447, 1430, 534, 1091, 1449]

TEST 2 — Decoder generation from sparse concepts

Middle tokens : 'the laboratoryThe scientist carefully conducted the experiment in the laboratoryThe scientist carefully conducted the'
Decoder output: ', and the number of people who have been killed'

Middle tokens : 'fresh ingredientsThe chef cooked a delicious meal with fresh ingredientsThe chef cooked a delicious'
Decoder output: ', and the use of the term "fear'

Middle tokens : 'government announced new policiesThe stock mar

In [ ]:
import matplotlib.pyplot as plt
import torch
import torch.optim as optim
from torch.utils.data import DataLoader

# ── Config ─────────────────────────────────────────────────────────────────
EPOCHS      = 5
BATCH_SIZE  = 16
NUM_SAMPLES = 10_000
LR          = 1e-4
LAYER_IDX   = 5

# ── Fresh init ─────────────────────────────────────────────────────────────
print("Rebuilding dataloader with 10k samples (this may take a minute)...")
gpt2_data   = prepare_pcd_data(model_name="gpt2", n_prefix=16, n_middle=16, n_suffix=16, num_samples=NUM_SAMPLES)
gpt2_loader = DataLoader(gpt2_data, batch_size=BATCH_SIZE, shuffle=True, collate_fn=collate_pcd)
print(f"✅ {len(gpt2_data)} samples loaded.")

# Re-initialise encoder and optimizer from scratch for a clean run
device = torch.device("cuda:0" if torch.cuda.is_available() else "cpu")
gpt2_encoder   = TopKEncoder(gpt2_d_model, gpt2_m_concepts, gpt2_k).to(device)
gpt2_encoder.train()

trainable_dec  = [p for p in gpt2_decoder.parameters() if p.requires_grad]
gpt2_optimizer = optim.AdamW(trainable_dec + list(gpt2_encoder.parameters()), lr=LR)

# ── Training loop with loss tracking ───────────────────────────────────────
history = {"ntp": [], "aux": [], "total": [], "epoch_avg_ntp": []}

gpt2_subject.eval()
gpt2_decoder.train()
gpt2_encoder.train()

capturer = ActivationCapture(gpt2_subject, layer_idx=LAYER_IDX)

for epoch in range(EPOCHS):
    epoch_ntp = []
    for step, batch in enumerate(gpt2_loader):
        gpt2_optimizer.zero_grad()

        prefix_ids = batch["prefix_ids"].to(device)
        middle_ids = batch["middle_ids"].to(device)
        suffix_ids = batch["suffix_ids"].to(device)

        with torch.no_grad():
            context_ids = torch.cat([prefix_ids, middle_ids], dim=1)
            gpt2_subject(input_ids=context_ids)
            a_mid = capturer.activations[:, prefix_ids.shape[1]:, :]

        encoded_a, sparse_latents = gpt2_encoder(a_mid)

        embed_layer   = gpt2_decoder.get_input_embeddings()
        suffix_embeds = embed_layer(suffix_ids).to(encoded_a.dtype)
        combined_embeds = torch.cat([encoded_a, suffix_embeds], dim=1)

        ignore_pad = torch.full(
            (suffix_ids.shape[0], encoded_a.shape[1]), -100,
            dtype=torch.long, device=device
        )
        labels = torch.cat([ignore_pad, suffix_ids], dim=1)

        outputs  = gpt2_decoder(inputs_embeds=combined_embeds, labels=labels)
        loss_ntp = outputs.loss
        loss_aux = compute_auxiliary_loss(gpt2_encoder, a_mid, sparse_latents)
        total    = loss_ntp + loss_aux

        total.backward()
        gpt2_optimizer.step()

        history["ntp"].append(loss_ntp.item())
        history["aux"].append(loss_aux.item())
        history["total"].append(total.item())
        epoch_ntp.append(loss_ntp.item())

    avg = sum(epoch_ntp) / len(epoch_ntp)
    history["epoch_avg_ntp"].append(avg)
    print(f"Epoch {epoch} | avg NTP: {avg:.4f} | steps: {len(epoch_ntp)}")

capturer.remove()
print("✅ Multi-epoch training complete.")

# ── Loss curve plot ─────────────────────────────────────────────────────────
fig, axes = plt.subplots(1, 2, figsize=(14, 4))
steps = range(len(history["ntp"]))

# Left: per-step NTP + smoothed trend
window = max(1, len(history["ntp"]) // 50)
smoothed = [
    sum(history["ntp"][max(0, i-window):i+1]) / len(history["ntp"][max(0, i-window):i+1])
    for i in steps
]
axes[0].plot(steps, history["ntp"], alpha=0.25, color="steelblue", label="NTP (raw)")
axes[0].plot(steps, smoothed,       color="steelblue", linewidth=2, label="NTP (smoothed)")
axes[0].set_title("NTP Loss per Step")
axes[0].set_xlabel("Step")
axes[0].set_ylabel("Loss")
axes[0].legend()

# Mark epoch boundaries
steps_per_epoch = len(history["ntp"]) // EPOCHS
for e in range(1, EPOCHS):
    axes[0].axvline(e * steps_per_epoch, color="gray", linestyle="--", alpha=0.5)

# Right: per-epoch average NTP
axes[1].plot(range(EPOCHS), history["epoch_avg_ntp"], marker="o", color="darkorange", linewidth=2)
axes[1].set_title("Average NTP Loss per Epoch")
axes[1].set_xlabel("Epoch")
axes[1].set_ylabel("Avg NTP Loss")
axes[1].set_xticks(range(EPOCHS))

plt.suptitle("GPT-2 PCD Stage 1 Training", fontsize=13, fontweight="bold")
plt.tight_layout()
plt.savefig("gpt2_pcd_loss_curve.png", dpi=150)
plt.show()
print("Plot saved to gpt2_pcd_loss_curve.png")

Rebuilding dataloader with 10k samples (this may take a minute)...
Loading tokenizer for gpt2...


Resolving data files:   0%|          | 0/2410 [00:00<?, ?it/s]

KeyboardInterrupt: 

In [ ]:
print("PyTorch version :", torch.__version__)
print("CUDA available  :", torch.cuda.is_available())
if torch.cuda.is_available():
    print("GPU             :", torch.cuda.get_device_name(0))
    print("VRAM            :", round(torch.cuda.get_device_properties(0).total_memory / 1e9, 1), "GB")
else:
    print("⚠️  No CUDA GPU detected — training will run on CPU and be very slow.")
    print("   Options:")
    print("   1. Run this notebook on Google Colab (free T4 GPU)")
    print("   2. Install CUDA-enabled PyTorch if you have an NVIDIA GPU:")
    print("      pip install torch torchvision torchaudio --index-url https://download.pytorch.org/whl/cu121")

print("\nCurrent training device:", torch.device("cuda:0" if torch.cuda.is_available() else "cpu"))

PyTorch version : 2.9.1+cpu
CUDA available  : False
⚠️  No CUDA GPU detected — training will run on CPU and be very slow.
   Options:
   1. Run this notebook on Google Colab (free T4 GPU)
   2. Install CUDA-enabled PyTorch if you have an NVIDIA GPU:
      pip install torch torchvision torchaudio --index-url https://download.pytorch.org/whl/cu121

Current training device: cpu


In [ ]:
print("PyTorch version :", torch.__version__)
print("CUDA available  :", torch.cuda.is_available())
if torch.cuda.is_available():
    print("GPU             :", torch.cuda.get_device_name(0))
    print("VRAM            :", round(torch.cuda.get_device_properties(0).total_memory / 1e9, 1), "GB")
else:
    print("⚠️  No CUDA GPU detected — training will run on CPU and be very slow.")
    print("   Options:")
    print("   1. Run this notebook on Google Colab (free T4 GPU)")
    print("   2. Install CUDA-enabled PyTorch if you have an NVIDIA GPU:")
    print("      pip install torch torchvision torchaudio --index-url https://download.pytorch.org/whl/cu121")

print("\nCurrent training device:", torch.device("cuda:0" if torch.cuda.is_available() else "cpu"))

PyTorch version : 2.9.1+cpu
CUDA available  : False
⚠️  No CUDA GPU detected — training will run on CPU and be very slow.
   Options:
   1. Run this notebook on Google Colab (free T4 GPU)
   2. Install CUDA-enabled PyTorch if you have an NVIDIA GPU:
      pip install torch torchvision torchaudio --index-url https://download.pytorch.org/whl/cu121

Current training device: cpu


# Llama-3.2-8B Run

Full PCD Stage 1 training on `meta-llama/Llama-3.2-8B`.

Key differences from the GPT-2 prototype:

| Setting | GPT-2 | Llama-3.2-8B |
|---|---|---|
| `d_model` | 768 | 4096 |
| `layer_idx` | 5 / 12 | 16 / 32 |
| `m_concepts` | 1536 | 8192 |
| `k` | 32 | 64 |
| `dtype` | float32 | bfloat16 (4-bit NF4 QLoRA) |
| LoRA targets | `c_attn` | `q_proj` `k_proj` `v_proj` `o_proj` |
| Segment length | 16/16/16 | 32/32/32 |
| Batch size | 16 | 4 (grad accum ×8 → effective 32) |

**VRAM requirement:** ~10–12 GB with 4-bit quantization (fits on A100 40 GB with room to spare; tight but possible on a 16 GB T4).

In [ ]:
import torch
from transformers import AutoModelForCausalLM, BitsAndBytesConfig

LLAMA_MODEL = "meta-llama/Llama-3.2-8B"

# 4-bit NF4 quantization — halves VRAM vs bfloat16 with minimal quality loss
bnb_config = BitsAndBytesConfig(
    load_in_4bit=True,
    bnb_4bit_quant_type="nf4",
    bnb_4bit_compute_dtype=torch.bfloat16,
    bnb_4bit_use_double_quant=True,   # nested quantization saves ~0.4 bits/param extra
)

print(f"Loading {LLAMA_MODEL} subject model (frozen, 4-bit NF4)...")
llama_subject = AutoModelForCausalLM.from_pretrained(
    LLAMA_MODEL,
    quantization_config=bnb_config,
    device_map="auto",
    torch_dtype=torch.bfloat16,
)
llama_subject.eval()
for p in llama_subject.parameters():
    p.requires_grad = False

# Report memory
if torch.cuda.is_available():
    used = torch.cuda.memory_allocated() / 1e9
    total = torch.cuda.get_device_properties(0).total_memory / 1e9
    print(f"✅ Subject model loaded.  GPU memory: {used:.1f} / {total:.1f} GB used.")
else:
    print("✅ Subject model loaded (CPU — training will be slow).")

In [ ]:
from peft import LoraConfig, get_peft_model, TaskType, prepare_model_for_kbit_training

print(f"Loading {LLAMA_MODEL} decoder (QLoRA)...")
llama_decoder_base = AutoModelForCausalLM.from_pretrained(
    LLAMA_MODEL,
    quantization_config=bnb_config,
    device_map="auto",
    torch_dtype=torch.bfloat16,
)

# Required before applying LoRA to a 4-bit model: casts layer norms to float32
# and enables gradient checkpointing to reduce activation memory
llama_decoder_base = prepare_model_for_kbit_training(
    llama_decoder_base, use_gradient_checkpointing=True
)

lora_config_llama = LoraConfig(
    r=16,
    lora_alpha=32,
    target_modules=["q_proj", "k_proj", "v_proj", "o_proj"],  # all attention projections
    lora_dropout=0.05,
    bias="none",
    task_type=TaskType.CAUSAL_LM,
)

llama_decoder = get_peft_model(llama_decoder_base, lora_config_llama)
llama_decoder.train()

print("✅ Llama decoder initialized with QLoRA.")
llama_decoder.print_trainable_parameters()

if torch.cuda.is_available():
    used = torch.cuda.memory_allocated() / 1e9
    total = torch.cuda.get_device_properties(0).total_memory / 1e9
    print(f"   GPU memory after both models: {used:.1f} / {total:.1f} GB")

In [ ]:
import torch.optim as optim
from torch.utils.data import DataLoader
from transformers import get_cosine_schedule_with_warmup

# ── Hyperparameters ────────────────────────────────────────────────────────
LLAMA_N_PREFIX   = 32
LLAMA_N_MIDDLE   = 32
LLAMA_N_SUFFIX   = 32
LLAMA_NUM_SAMPLES = 50_000
LLAMA_BATCH_SIZE  = 4
LLAMA_GRAD_ACCUM  = 8        # effective batch = 32
LLAMA_EPOCHS      = 1
LLAMA_LR          = 1e-4
LLAMA_LAYER_IDX   = 16       # mid-network (layer 16 of 32)

# ── Dataset ────────────────────────────────────────────────────────────────
print(f"Preparing data ({LLAMA_NUM_SAMPLES} samples) with Llama tokenizer...")
llama_data = prepare_pcd_data(
    model_name=LLAMA_MODEL,
    n_prefix=LLAMA_N_PREFIX,
    n_middle=LLAMA_N_MIDDLE,
    n_suffix=LLAMA_N_SUFFIX,
    num_samples=LLAMA_NUM_SAMPLES,
)
llama_loader = DataLoader(
    llama_data, batch_size=LLAMA_BATCH_SIZE, shuffle=True, collate_fn=collate_pcd
)
print(f"✅ {len(llama_data)} samples, {len(llama_loader)} batches/epoch.")

# ── Encoder ────────────────────────────────────────────────────────────────
llama_d_model    = llama_subject.config.hidden_size  # 4096
llama_m_concepts = 8192
llama_k          = 64

enc_device = torch.device("cuda:0" if torch.cuda.is_available() else "cpu")
llama_encoder = TopKEncoder(llama_d_model, llama_m_concepts, llama_k)
llama_encoder = llama_encoder.to(enc_device).to(torch.bfloat16)
llama_encoder.train()
print(f"✅ Encoder: d_model={llama_d_model}, m_concepts={llama_m_concepts}, k={llama_k}")

# ── Optimizer + LR scheduler ───────────────────────────────────────────────
llama_trainable_dec = [p for p in llama_decoder.parameters() if p.requires_grad]
all_params = llama_trainable_dec + list(llama_encoder.parameters())

llama_optimizer = optim.AdamW(all_params, lr=LLAMA_LR, weight_decay=0.01)

total_opt_steps = (len(llama_loader) // LLAMA_GRAD_ACCUM) * LLAMA_EPOCHS
warmup_steps    = max(1, total_opt_steps // 10)
llama_scheduler = get_cosine_schedule_with_warmup(
    llama_optimizer, num_warmup_steps=warmup_steps, num_training_steps=total_opt_steps
)
print(f"✅ Optimizer ready. {total_opt_steps} optimizer steps, {warmup_steps} warmup steps.")

In [ ]:
import os, math
import matplotlib.pyplot as plt

CHECKPOINT_DIR  = "llama_pcd_checkpoints"
SAVE_EVERY_STEPS = 500   # optimizer steps between checkpoint saves
LOG_EVERY        = 10    # gradient-accumulation steps between console prints

os.makedirs(CHECKPOINT_DIR, exist_ok=True)

llama_subject.eval()
llama_decoder.train()
llama_encoder.train()

llama_capturer = ActivationCapture(llama_subject, layer_idx=LLAMA_LAYER_IDX)

history_llama = {"ntp": [], "aux": [], "total": [], "epoch_avg": []}
opt_step = 0

print(f"Starting Llama-3.2-8B PCD Stage 1 Training")
print(f"  {LLAMA_EPOCHS} epoch(s) | {len(llama_loader)} batches | grad accum {LLAMA_GRAD_ACCUM} → {total_opt_steps} optimizer steps")

try:
    for epoch in range(LLAMA_EPOCHS):
        epoch_ntp = []
        llama_optimizer.zero_grad()

        for step, batch in enumerate(llama_loader):
            prefix_ids = batch["prefix_ids"].to(enc_device)
            middle_ids = batch["middle_ids"].to(enc_device)
            suffix_ids = batch["suffix_ids"].to(enc_device)

            # 1. Frozen subject model — extract mid-layer activations
            with torch.no_grad():
                context_ids = torch.cat([prefix_ids, middle_ids], dim=1)
                llama_subject(input_ids=context_ids)
                a_mid = llama_capturer.activations[:, prefix_ids.shape[1]:, :].to(enc_device)

            # 2. Encode into sparse concepts
            encoded_a, sparse_latents = llama_encoder(a_mid)

            # 3. Decoder forward pass
            embed_layer   = llama_decoder.get_input_embeddings()
            suffix_embeds = embed_layer(suffix_ids).to(encoded_a.dtype)
            combined_embeds = torch.cat([encoded_a, suffix_embeds], dim=1)

            ignore_pad = torch.full(
                (suffix_ids.shape[0], encoded_a.shape[1]), -100,
                dtype=torch.long, device=enc_device
            )
            labels  = torch.cat([ignore_pad, suffix_ids], dim=1)
            outputs = llama_decoder(inputs_embeds=combined_embeds, labels=labels)

            loss_ntp = outputs.loss
            loss_aux = compute_auxiliary_loss(llama_encoder, a_mid, sparse_latents)
            total    = (loss_ntp + loss_aux) / LLAMA_GRAD_ACCUM  # scale for accumulation

            total.backward()

            # Track unscaled losses for logging
            history_llama["ntp"].append(loss_ntp.item())
            history_llama["aux"].append(loss_aux.item())
            history_llama["total"].append((loss_ntp + loss_aux).item())
            epoch_ntp.append(loss_ntp.item())

            # Optimizer step every GRAD_ACCUM batches
            if (step + 1) % LLAMA_GRAD_ACCUM == 0:
                torch.nn.utils.clip_grad_norm_(all_params, max_norm=1.0)
                llama_optimizer.step()
                llama_scheduler.step()
                llama_optimizer.zero_grad()
                opt_step += 1

                if opt_step % LOG_EVERY == 0:
                    lr_now = llama_scheduler.get_last_lr()[0]
                    print(f"Epoch {epoch} | OptStep {opt_step}/{total_opt_steps} "
                          f"| NTP: {loss_ntp.item():.4f} | Aux: {loss_aux.item():.6f} "
                          f"| LR: {lr_now:.2e}")

                # Checkpoint
                if opt_step % SAVE_EVERY_STEPS == 0:
                    ckpt_path = os.path.join(CHECKPOINT_DIR, f"step_{opt_step}")
                    llama_decoder.save_pretrained(ckpt_path)
                    torch.save(llama_encoder.state_dict(),
                               os.path.join(ckpt_path, "encoder.pt"))
                    print(f"  💾 Checkpoint saved → {ckpt_path}")

        avg = sum(epoch_ntp) / len(epoch_ntp)
        history_llama["epoch_avg"].append(avg)
        print(f"\n✅ Epoch {epoch} done | avg NTP: {avg:.4f}\n")

except Exception as e:
    import traceback
    print(f"❌ Training failed: {e}")
    traceback.print_exc()
finally:
    llama_capturer.remove()

# ── Final checkpoint ───────────────────────────────────────────────────────
final_path = os.path.join(CHECKPOINT_DIR, "final")
llama_decoder.save_pretrained(final_path)
torch.save(llama_encoder.state_dict(), os.path.join(final_path, "encoder.pt"))
print(f"💾 Final checkpoint saved → {final_path}")

# ── Loss curve ─────────────────────────────────────────────────────────────
if history_llama["ntp"]:
    fig, ax = plt.subplots(figsize=(12, 4))
    raw = history_llama["ntp"]
    window = max(1, len(raw) // 60)
    smoothed = [sum(raw[max(0,i-window):i+1]) / len(raw[max(0,i-window):i+1])
                for i in range(len(raw))]
    ax.plot(raw,      alpha=0.2, color="steelblue", label="NTP (raw)")
    ax.plot(smoothed, color="steelblue", linewidth=2, label="NTP (smoothed)")
    ax.set_title("Llama-3.2-8B PCD Stage 1 — NTP Loss", fontsize=13)
    ax.set_xlabel("Batch step")
    ax.set_ylabel("Loss")
    ax.legend()
    plt.tight_layout()
    plt.savefig("llama_pcd_loss_curve.png", dpi=150)
    plt.show()
    print("Plot saved to llama_pcd_loss_curve.png")